### DISCUSSION TITLE
The key / keybind budget

### DISCUSSION THREAD

**Discussion by LevFendi** [Factorio-Access/FactorioAccess#154] [General]
A topic that has rightfully been popping for the past few weeks is the key budget. This page offers a recap and a persistent place to keep discussing it.

Factorio lets you do a lot of things, and so there are several GUI menus and several keybinds for all kinds of things. The Access mod has substituted a lot of these GUI features with simple menus and additional keybinds. Other keybinds are being used for unique features of this mod. This has worked well until recently.

We have now reached a point where we have two issues: A. There are so many keys that it is becoming difficult to remember what to use where. And B. There are still things that we want to add and it is difficult to find the right keybinds for them.

For issue A, a solution being considered is the ability to press a key and get the context-specific list of available controls, similar to typing "help" to a console program. This is issue #94.

For both issues, a solution pattern being considered is to convert some of the keybind-based functions into menu-based functions. This is most evident for logistic request management. This is issue #88. 

For some non-gameplay menus that will be added in the future, such as the acheivements menu and the preferences menu, my plan is to have a single top-level mod menu where you can access sub-menus. For example, you will press "M" and this will open all those menus.

For some gameplay menus such as the warnings menu and the future production info menu, perhaps those can also be united under a main menu.

**Comment by ahicks92** [comment]
Didn't realize we were using this as well as Discord till the other day

We could use more [] because alt is still available.  Anonoying but also not--it's amazing how well wasd and [] work out for speed.

Something that kind of gets...it's not so obvious to the sighted I guess. But GUIs can be good.  I do think that (after my WIP work) we should discuss good GUI design and see what falls out.  As one perhaps counterintuitive example, the fast travel menu's horizontal "commands" is actually very nice and has in fact been done before for that reason (and indeed I am taking that into account).  E.G. if you remember the position then the player can just move up and down and click to do the same action in (relative) bulk (or just make it a proper table--we do need one of those, no reason some cells can't be clickable.  Same to the user but probably a better abstraction).

One place I'll advocate hard for that for example--it'd make navigating circuit menus literally 3 or 4 times faster

**Reply by LevFendi** [reply]
(I accidentally made a thread instead of a general reply)

**Comment by LevFendi** [comment]
As a quick note here:
CONTROL+W/S can be remapped to ALT+W/S without overwriting anything. This would allow us to freely use CONTROL+W/A/S/D. What to use it for is a separate discussion...
